File: samples/snes/make_cart.go
package main

import "fmt"
import "os"

func main() {
  if len(os.Args) != 3 {
    fmt.Println("Usage: " + os.Args[0] + " <binfile> <smcfile>")
    os.Exit(0)
  }

  file_in, err := os.Open(os.Args[1])

  if err != nil {
    panic(err)
  }

  defer file_in.Close()

  stat, err := file_in.Stat()
  if err != nil {
    panic(err)
  }

  size := int(stat.Size())
  cart := make([]byte, size)
  header := make([]byte, 512)

  file_in.Read(cart)
  file_in.Close()

  if (len(cart) % 8192) != 0 {
    fmt.Println("Error: ROM not a multiple of 8k")
    return
  }

  i := len(cart) / 8192

  header[0] = byte(i & 0xff)
  header[1] = byte(i >> 8)

  file_out, err := os.Create(os.Args[2])
  file_out.Write(header)
  file_out.Write(cart)
  file_out.Close()

}

File: samples/snes/make_cart.py
#!/usr/bin/env python3

import sys

if len(sys.argv) != 3:
  print("Usage: python3 make_cart.py <binfile> <smcfile>")
  sys.exit(0)

fp = open(sys.argv[1], "rb")
cart = bytearray(fp.read())
fp.close()

if (len(cart) % 8192) != 0:
  print("Error: ROM not a multiple of 8k")
  sys.exit(1)

# FIXME: This checksum calculation doesn't seem to work.
checksum = 0

for i in range(0, len(cart)):
  checksum += cart[i]

checksum &= 0xffff

cart[0xffde - 32768] = checksum & 0xff
cart[0xffdf - 32768] = checksum >> 8
checksum = checksum ^ 0xffff
cart[0xffdc - 32768] = checksum & 0xff
cart[0xffdd - 32768] = checksum >> 8

header = bytearray(512)
blocks = int(len(cart) / 8192)

header[0] = blocks & 0xff
header[1] = blocks >> 8

fp = open(sys.argv[2], "wb")
fp.write(header)
fp.write(cart)
fp.close()

